cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -185,6 +190,12 @@ where
         shuffling_seed: Randomness,
         maybe_new_runtime: Option<Cow<'static, [u8]>>,
     ) -> Result<(), sp_blockchain::Error> {
+        let fraud_proof_permit = self.unsigned_submitter.try_reserve().map_err(|_| {
+            sp_blockchain::Error::Application(Box::from(
+                "Channel for submitting unsigned extrinsics is full",
+            ))
+        })?;
```

Comment:
Previous revision was returning `Ok` when channel is full, so this change isn't equivalent. Will some bundles be skipped entirely if this error happens and break the client? This is what seems to happen judging by the place where `process_bundles` is called from, I don't see any retry mechanism or anything like that, it just moves on with broken state.

Reply:
The only difference from the previous version is that an error message of the channel being full will be logged, but now I realize that none of them is appropriate without the retry mechanism. The state won't be broken as the program now will panic if a primary block is skipped and the bundle processor attempts to process the next primary block due to the primary number check(see the assertation below `New secondary best number must be equal to the primary number`).

Okay, now the problem becomes more complicated, currently, we rely on the primary block import notification which is sort of one-off, for each imported primary block, the notification of that block will only be sent once and gone. When processing a primary block, if `process_bundles` returns an error, the executor is just broken at present, the state is not corrupted, but it can not move forward until the error is fixed and restarted.

Besides the retry mechanism, another problem begins to emerge: even though the retry mechanism exists, while the executor is retrying, the primary node is still syncing, which is a full node instead of an archival node and we won't expect the executor has to maintain a primary node keeping all the blocks, so if the retry takes too long, the primary node can proceed to a much higher block and prune the old blocks which are not yet processed by the executor, even when the executor retry is successful later, the corresponding primary block is unavailable and the executor get stuck there. I need some more time to think it over.

Reply:
Rebased and added TODO which will be worked on in the nearest feature.